File: android/app/src/main/kotlin/hk/smsguard/app/rules/Detector.kt
package hk.smsguard.app.rules

interface BlocklistLookup {
    fun isBlockedUrlCanonical(canonical: String): Boolean
    fun isBlockedDomain(registrableDomain: String): Boolean

    companion object {
        val EMPTY: BlocklistLookup = object : BlocklistLookup {
            override fun isBlockedUrlCanonical(canonical: String) = false
            override fun isBlockedDomain(registrableDomain: String) = false
        }
    }
}

data class DetectorContext(
    val registry: SsrsRegistry,
    val blocklist: BlocklistLookup,
)

data class DetectionDetail(
    val verdict: Verdict,
    val extractedUrls: List<ExtractedUrl>,
    val contentHits: List<ContentRuleHit>,
)

fun detect(sms: IncomingSms, ctx: DetectorContext): DetectionDetail {
    val ssrsOutcome = checkSsrs(sms, ctx.registry)
    val ssrsVerdict = ssrsCheckToVerdict(ssrsOutcome)

    if (ssrsVerdict != null && ssrsVerdict.label == VerdictLabel.TRUSTED) {
        return DetectionDetail(verdict = ssrsVerdict, extractedUrls = emptyList(), contentHits = emptyList())
    }
    if (ssrsVerdict != null && ssrsVerdict.label == VerdictLabel.HIGH_CONFIDENCE_PHISHING) {
        return DetectionDetail(
            verdict = ssrsVerdict,
            extractedUrls = extractUrls(sms.body),
            contentHits = findContentRuleHits(sms.body),
        )
    }

    val urls = extractUrls(sms.body)
    val blockedHits = mutableListOf<String>()
    for (u in urls) {
        when {
            ctx.blocklist.isBlockedUrlCanonical(u.canonical) -> blockedHits += "blocklist.url:${u.canonical}"
            ctx.blocklist.isBlockedDomain(u.registrableDomain) -> blockedHits += "blocklist.domain:${u.registrableDomain}"
        }
    }

    val contentHits = findContentRuleHits(sms.body)

    val fired = mutableListOf<String>()
    val signalScores = mutableListOf<Double>()
    if (ssrsVerdict != null) {
        fired += ssrsVerdict.firedRuleIds
        signalScores += ssrsVerdict.score
    }
    for (hit in blockedHits) {
        fired += hit
        signalScores += 0.9
    }
    for (hit in contentHits) {
        fired += "content.${hit.ruleId}"
        signalScores += severityToScore(hit.severity)
    }

    val score = noisyOr(signalScores)
    val label = labelFromScore(score)
    val verdict = Verdict(
        label = label,
        score = score,
        firedRuleIds = fired,
        explanationKey = pickExplanationKey(label, blockedHits, contentHits, ssrsVerdict),
        explanationParams = emptyMap(),
    )
    return DetectionDetail(verdict = verdict, extractedUrls = urls, contentHits = contentHits)
}

private fun severityToScore(s: Severity): Double = when (s) {
    Severity.HIGH -> 0.7
    Severity.MEDIUM -> 0.4
    Severity.LOW -> 0.2
}

private fun noisyOr(scores: List<Double>): Double {
    if (scores.isEmpty()) return 0.0
    var prod = 1.0
    for (s in scores) {
        val clamped = s.coerceIn(0.0, 1.0)
        prod *= 1.0 - clamped
    }
    return 1.0 - prod
}

private fun labelFromScore(score: Double): VerdictLabel = when {
    score >= 0.9 -> VerdictLabel.HIGH_CONFIDENCE_PHISHING
    score >= 0.7 -> VerdictLabel.LIKELY_SCAM
    score >= 0.4 -> VerdictLabel.SUSPICIOUS
    else -> VerdictLabel.NO_SIGNAL
}

private fun pickExplanationKey(
    label: VerdictLabel,
    blockedHits: List<String>,
    contentHits: List<ContentRuleHit>,
    ssrsVerdict: Verdict?,
): String {
    if (ssrsVerdict != null && ssrsVerdict.label != VerdictLabel.NO_SIGNAL) return ssrsVerdict.explanationKey
    if (blockedHits.isNotEmpty()) return "blocklist.url_or_domain"
    if (contentHits.isNotEmpty()) return "content.${contentHits[0].category.name.lowercase()}"
    return "verdict.${label.name.lowercase()}"
}

File: shared/rules/src/detector.ts
import { findContentRuleHits } from './content-rules.js';
import { checkSsrs, ssrsCheckToVerdict } from './ssrs-verifier.js';
import type {
  ContentRuleHit,
  ExtractedUrl,
  IncomingSms,
  SsrsRegistry,
  Verdict,
  VerdictLabel,
} from './types.js';
import { extractUrls } from './url-extractor.js';

export interface BlocklistLookup {
  isBlockedUrlCanonical(canonical: string): boolean;
  isBlockedDomain(registrableDomain: string): boolean;
}

export interface DetectorContext {
  readonly registry: SsrsRegistry;
  readonly blocklist: BlocklistLookup;
}

export interface DetectionDetail {
  readonly verdict: Verdict;
  readonly extractedUrls: readonly ExtractedUrl[];
  readonly contentHits: readonly ContentRuleHit[];
}

export function detect(sms: IncomingSms, ctx: DetectorContext): DetectionDetail {
  const ssrsOutcome = checkSsrs(sms, ctx.registry);
  const ssrsVerdict = ssrsCheckToVerdict(ssrsOutcome);

  if (ssrsVerdict !== null && ssrsVerdict.label === 'trusted') {
    return { verdict: ssrsVerdict, extractedUrls: [], contentHits: [] };
  }

  // Carrier-verified #prefix that we don't recognize: trust the OFCA
  // registration. Content rules don't run for #prefix senders because the
  // whole SSRS scheme is "if you see #, the carrier already verified the
  // org behind it." Marketing copy from real registered orgs frequently
  // contains urgency/credential phrasing that would otherwise flag.
  if (ssrsOutcome.outcome === 'unknown_hash_prefix') {
    return {
      verdict: {
        label: 'no_signal',
        score: 0.0,
        firedRuleIds: [`ssrs.carrier_verified:${ssrsOutcome.observedPrefix ?? ''}`],
        explanationKey: 'ssrs.carrier_verified',
        explanationParams: { prefix: ssrsOutcome.observedPrefix ?? '' },
      },
      extractedUrls: [],
      contentHits: [],
    };
  }

  const urls = extractUrls(sms.body);
  const blockedHits: string[] = [];
  for (const u of urls) {
    if (ctx.blocklist.isBlockedUrlCanonical(u.canonical)) {
      blockedHits.push(`blocklist.url:${u.canonical}`);
    } else if (ctx.blocklist.isBlockedDomain(u.registrableDomain)) {
      blockedHits.push(`blocklist.domain:${u.registrableDomain}`);
    }
  }

  const contentHits = findContentRuleHits(sms.body);

  const fired: string[] = [];
  const signalScores: number[] = [];
  if (ssrsVerdict !== null) {
    fired.push(...ssrsVerdict.firedRuleIds);
    signalScores.push(ssrsVerdict.score);
  }
  // Org claim without prefix: a contributing signal whose weight depends on
  // whether the body also contains a URL. Plain mentions of HSBC/匯豐 in
  // marketing chatter shouldn't tip the verdict on their own; but org-name
  // at message start + a URL is a strong impersonation pattern.
  if (ssrsOutcome.outcome === 'phishing_claims_org_without_prefix') {
    if (urls.length > 0) {
      fired.push('ssrs.org_claimed_without_prefix_with_url');
      signalScores.push(0.85);
    } else {
      fired.push('ssrs.org_claimed_without_prefix');
      signalScores.push(0.3);
    }
  }
  for (const hit of blockedHits) {
    fired.push(hit);
    signalScores.push(0.9);
  }
  for (const hit of contentHits) {
    fired.push(`content.${hit.ruleId}`);
    signalScores.push(severityToScore(hit.severity));
  }

  const score = noisyOr(signalScores);

  const label = labelFromScore(score);
  const verdict: Verdict = {
    label,
    score,
    firedRuleIds: fired,
    explanationKey: pickExplanationKey(label, blockedHits, contentHits, ssrsVerdict, fired),
    explanationParams: {},
  };
  return { verdict, extractedUrls: urls, contentHits };
}

function severityToScore(s: 'high' | 'medium' | 'low'): number {
  if (s === 'high') return 0.7;
  if (s === 'medium') return 0.4;
  return 0.2;
}

function noisyOr(scores: readonly number[]): number {
  if (scores.length === 0) return 0;
  let prod = 1;
  for (const s of scores) {
    const clamped = Math.max(0, Math.min(1, s));
    prod *= 1 - clamped;
  }
  return 1 - prod;
}

function labelFromScore(score: number): VerdictLabel {
  if (score >= 0.9) return 'high_confidence_phishing';
  if (score >= 0.7) return 'likely_scam';
  if (score >= 0.4) return 'suspicious';
  return 'no_signal';
}

function pickExplanationKey(
  label: VerdictLabel,
  blockedHits: readonly string[],
  contentHits: readonly ContentRuleHit[],
  ssrsVerdict: Verdict | null,
  fired: readonly string[],
): string {
  if (ssrsVerdict !== null && ssrsVerdict.label !== 'no_signal') {
    return ssrsVerdict.explanationKey;
  }
  if (fired.some((f) => f.startsWith('ssrs.org_claimed_without_prefix'))) {
    return 'ssrs.org_claimed_without_prefix';
  }
  if (blockedHits.length > 0) return 'blocklist.url_or_domain';
  if (contentHits.length > 0) return `content.${contentHits[0]!.category}`;
  return `verdict.${label}`;
}
